### ISSUE TITLE
[ENH] Provide a "compatibility over performance" C compile mode

### ISSUE THREAD

**Issue by scoder** [cython/cython#5249]
### Is your feature request related to a problem? Please describe.

Many Cython users would benefit from a compile mode (read, C macro) that increases the compatibility with future CPython releases, even at the cost of some performance.

The code that Cython generates for CPython takes advantage of several implementation details, e.g. PyLong digits or PyList/PyTuple arrays. These can change in new CPython releases, which requires Cython to adapt and users to regenerate their code.

In many cases, users care more about a future proof extension module than a highly performant one. The Limited C-API in CPython was introduced in order to cater for this use case, but is not currently complete enough to implement all of Cython (really, all of Python) on top of it.

Related discussions:
- the "LIMITED_API" option that replaces all usages of CPython internals with generic, public C-API calls (https://github.com/cython/cython/issues/2542)
- the ticket that @markshannon opened, asking us to avoid implementation details (https://github.com/cython/cython/issues/4382)
- the list of CPython internals that we use (https://github.com/cython/cython/issues/4635)
- several incompatibilities between the code that Cython generated and changes that appeared in CPython 3.12a5 (https://github.com/cython/cython/issues/5238)

### Describe the solution you'd like.

There should be a separate compile option, similar to `CYTHON_LIMITED_API` (which enables the `CYTHON_COMPILING_IN_LIMITED_API` set of feature flags), but without the limitations of the Limited C-API.

### Describe alternatives you've considered.

We could disable all usages of internals for future CPython releases, and only enable their usage in newer Cython releases that were tested on the respective CPython versions. This would make sure that Cython generated code keeps working in future CPython releases (as long as there are no changes to existing public C-API functions), but would introduce an enforced slow-down in new CPython releases that would require users to regenerate their code with a more recent Cython release to solve.

### Additional context

One disadvantage is that adding a new C compile option increases the number of CI jobs that are necessary to guarantee the correct functioning of the generated code.

**Comment by tacaswell** [comment]
> These can change in new CPython releases, which requires Cython to adapt

This seems like a real cost to me

> and users to regenerate their code.

but this does not.  Given that cython is pip installable (and the improvements for specifying build-time requirements in the broader Python ecosystem) I think is completely reasonable to expect users to re-cythonize when installing a package from source.  If you are building from source then you also need a c-compiler which is already much harder.  I think that cython should take the view that direct users of cython are sophisticated enough to have no problems installing cython and the 2nd-level users (users of projects that use cython) are either able easily get cython installed or are consuming some sort of binary build (wheels, conda packages, system packages, ....) where someone else is responsible for building.


If the goal is to make the path to supporting new versions of CPython smoother (it goes slower rather than fails so you get the bug report after CPython is released and someone's benchmarks got slower rather than me complaining as soon as it lands on CPython main) and that make the lives of cython developers easier then this seems like a great idea!

If the primary motivation is to make it so that down-stream users have to re-cythonize less then this seems like a lot of work for no practical benefit.

**Comment by dalcinl** [comment]
@tacaswell I'm not sure what is the official recommendation now, but years ago Cython told folks to ship generated C sources within sdists. I've been doing so with my project until last release, but for the future I'll add Cython as a build-time dependency to run the Cython -> C translation at build time, precisely to avoid the issues we are discussing in here. 

That being said, we still have the problem of working with in-development CPython versions. Recently, GitHub's `actions/setup-python` upgraded its `Py312-dev` to a beta with breaking changes. Cython-generated C code was no longer able to compile, my CI workflow got disrupted. I ended up doing this in my `setup.py`:

```python
# setup.py
maxknow_python = (3, 11)
...
def get_extension(...):
    ext = ...
    if sys.version_info[:2] > maxknow_python:
        ext['define_macros'].extend([
            ('CYTHON_FAST_PYCALL', 0),
            ('CYTHON_FAST_THREAD_STATE', 0),
            ('CYTHON_USE_DICT_VERSIONS', 0),
            ('CYTHON_USE_PYLONG_INTERNALS', 0),
            ('CYTHON_USE_PYLIST_INTERNALS', 0),
            ('CYTHON_USE_UNICODE_INTERNALS', 0),
        ])
```
and I am still not sure if I turned off all the optimizations that may break in the next Py312 API change. 
It would be much easier to have a single define (e.g. `CYTHON_USE_CPYTHON_INTERNALS`) to set.

PS: After writing this comment, I realized I did it slightly wrong. I should define `maxknow_python` to the latest Py312 beta that is know to work.

**Comment by scoder** [comment]
> Cython told folks to ship generated C sources within sdists

Separate issue: https://github.com/cython/cython/issues/5089

**Comment by scoder** [comment]
> After writing this comment, I realized I did it slightly wrong. I should define `maxknow_python` to the latest Py312 beta that is know to work.

Right. As it stands, your version guard would already apply to Python 3.11.1, not just 3.12.0a5, which introduced several breaking changes.

**Comment by dalcinl** [comment]
> your version guard would already apply to Python 3.11.1, not just 3.12.0a5

Well, note the `[:2]` slice after `std.version_info`. But still, 3.12.0a4 was good, so I should do better.

**Comment by markshannon** [comment]
What would be even better, would be an option to choose between cross-version compatibility and performance when compiling the C code generated by Cython.
That will need CPython support, so that we can provide the inline functions needed for the performance mode.

Is there a list anywhere of all the inline functions that Cython defines that access CPython internals?

**Comment by scoder** [comment]
> What would be even better, would be an option to choose between cross-version compatibility and performance when compiling the C code generated by Cython.

That was the idea of the Limited C-API. Cython has partial support for that (at C compile time, see https://github.com/cython/cython/issues/2542), which should be extended (both in Cython and CPython).

However, I'm not sure that that's completely identical to the use case in this ticket. The described C compile mode could make use of more of the C-API than the Limited API, while still guarding a reasonably high level of forward compatibility.

Basically, if the Limited API was complete enough to limit Cython to it (read: to implement Python on top of it it), we could use the existing "CYTHON_LIMITED_API" compile mode. But it's not, at least not as it stands. We may eventually end up merging the two modes at some point, but we need something complete for the time being. Something that users can explicitly enable and still get what they meant in the future.

**Comment by scoder** [comment]
> Is there a list anywhere of all the inline functions that Cython defines that access CPython internals?

There's https://github.com/cython/cython/issues/4635 which lists the internals that we use. Regarding the usages of these internals in the code, the best way to find them is probably to search for the respective macro guards in the sources, as listed above in https://github.com/cython/cython/issues/5249#issuecomment-1431631813 and (completely) near the top of https://github.com/cython/cython/blob/master/Cython/Utility/ModuleSetupCode.c.

Several of these occurrences are not easy to extract, though, (which is what you seem to be after). Being a code generator, Cython can afford to adapt the usage to the source code that it translates. So there's a mix of pre-written utility functions, template-generated utility functions, and completely generated code snippets where the guards are being used.

Also note that there are different reasons why we implement our own helper functions. One is performance, another is incomplete coverage of Python's semantics in CPython's C-API, sometimes we really know better because Cython has C/Python type information, and sometimes it's just "because we can afford to do so in a code generator" (e.g. the generated code can be better adapted to the use case than a generic C-API call).

**Comment by da-woods** [comment]
As far as is possible I think we should keep this to just being a convenient way of defining all the other compatibility macros. Partly because I think that practically gets pretty close already, and partly as a matter of principle to try to minimize the amount of work and extra testing that's involved in this.
